homeassistant/helpers/unit_system.py

```diff
@@ -0,0 +1,97 @@
+"""Unit system helper class and methods."""
+
+import logging
+from homeassistant.const import (
+    TEMP_CELSIUS, TEMP_FAHRENHEIT, LENGTH_CENTIMETERS, LENGTH_METERS,
+    LENGTH_KILOMETERS, LENGTH_INCHES, LENGTH_FEET, LENGTH_YARD, LENGTH_MILES,
+    VOLUME_LITERS, VOLUME_MILLILITERS, VOLUME_GALLONS, VOLUME_FLUID_OUNCE,
+    MASS_GRAMS, MASS_KILOGRAMS, MASS_OUNCES, MASS_POUNDS, SYSTEM_METRIC,
+    SYSTEM_IMPERIAL)
+
+_LOGGER = logging.getLogger(__name__)
+
+TYPE_LENGTH = 'length'  # type: str
+TYPE_MASS = 'mass'  # type: str
+TYPE_VOLUME = 'volume'  # type: str
+TYPE_TEMPERATURE = 'temperature'  # type: str
+
+LENGTH_UNITS = [
+    LENGTH_MILES,
+    LENGTH_YARD,
+    LENGTH_FEET,
+    LENGTH_INCHES,
+    LENGTH_KILOMETERS,
+    LENGTH_METERS,
+    LENGTH_CENTIMETERS,
+]
+
+MASS_UNITS = [
+    MASS_POUNDS,
+    MASS_OUNCES,
+    MASS_KILOGRAMS,
+    MASS_GRAMS,
+]
+
+VOLUME_UNITS = [
+    VOLUME_GALLONS,
+    VOLUME_FLUID_OUNCE,
+    VOLUME_LITERS,
+    VOLUME_MILLILITERS,
+]
+
+TEMPERATURE_UNITS = [
+    TEMP_FAHRENHEIT,
+    TEMP_CELSIUS,
+]
+
+NOT_RECOGNIZED_TEMPLATE = '{} is not a recognized {} unit.'  # type: str
+
+
+def is_valid_unit(unit: str, unit_type: str) -> bool:
+    """Check if the unit is valid for it's type."""
+    if unit_type == TYPE_LENGTH:
+        units = LENGTH_UNITS
+    elif unit_type == TYPE_TEMPERATURE:
+        units = TEMPERATURE_UNITS
+    elif unit_type == TYPE_MASS:
+        units = MASS_UNITS
+    elif unit_type == TYPE_VOLUME:
+        units = VOLUME_UNITS
+    else:
+        return False
+
+    return unit in units
+
+
+class UnitSystem(dict):
```

Comment:
Is this common for serializing custom objects?

On Aug 2, 2016 1:13 AM, "Paulus Schoutsen" <notifications@github.com> wrote:

> In homeassistant/helpers/unit_system.py
> <https://github.com/home-assistant/home-assistant/pull/2705#discussion_r73104307>
> :
>
> > +    """Check if the unit is valid for it's type."""
> > +    if unit_type == TYPE_LENGTH:
> > +        units = LENGTH_UNITS
> > +    elif unit_type == TYPE_TEMPERATURE:
> > +        units = TEMPERATURE_UNITS
> > +    elif unit_type == TYPE_MASS:
> > +        units = MASS_UNITS
> > +    elif unit_type == TYPE_VOLUME:
> > +        units = VOLUME_UNITS
> > +    else:
> > +        return False
> > +
> > +    return unit in units
> > +
> > +
> > +class UnitSystem(dict):
>
> For the new UnitSystem object, also add an as_dict() method for sending
> it to the frontend. See example in Event.
> <https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/core.py#L204-L211>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/2705/files/ca22ff2a9e47bec2130f03efd745043aaaf76993#r73104307>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC2fZWwGxq3xh4kRCnk5lnkFUKBGgB-rks5qbu4kgaJpZM4JaO1y>
> .
>